Symmetry-Breaking Findif, and Reference First

JonathonMisiewicz opened the pull request:
## Description
Fixes #1052 at long, long last: when the user explicitly declares a symmetry, findif is now smart enough to put displacements in the proper subgroup.

As part of the fix, control over molecule cloning for finite difference has passed to `_process_displacement`. It seemed easier (and more appropriate) than adding code to make a new clone in each for loop. The clone would be necessary so that each displacement can have its own symmetry.

Since I was touching that part of the code anyways, I have reorganized so that the reference is guaranteed to be the first displacement. This has the advantage that if orbital reading is enabled, the following displacements will be able to read from the reference, as intended.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Fixes #1052
- [x] Exposes more C functions to Python
- [x] Move the reference geometry first for ease of orbital reads

## Checklist
- [x] Tests modified to account for bug fix
- [x] All findif tests passing

## Status
- [x] Ready for review
- [x] Ready for merge

dgasmith commented on psi4/driver/driver.py:
```diff
@@ -1269,13 +1275,8 @@ def hessian(name, **kwargs):
-        findif_meta_dict = driver_findif.hessian_from_gradient_geometries(moleculeclone, irrep)
-        moleculeclone.reinterpret_coordentry(False)
-        moleculeclone.fix_orientation(True)
+        findif_meta_dict = driver_findif.hessian_from_gradient_geometries(molecule, irrep)
```
Should this be `clone` here?

JonathonMisiewicz replied:
No.

None of the geometry generators change the molecule, so there's no need to have a clone for purposes of generating the metadict. Cloning is only needed for the computations and is now done in `_process_displacement`, because different displacements need different clones for symmetry purposes.

andysim commented:
This looks great!  Thanks for figuring out this very tricky problem.  A long time ago, @fevangelista and I figured out the machinery to map symmetry-dependent arrays like `DOCC` and `SOCC` into the appropriate subgroups.  Would you mind quickly testing to see if `DOCC` can be specified in c2v symmetry with your current version?  If not, we'll have to open a follow up PR to make sure this mapping is done (I think it's best left out of this PR to avoid scope creep).

dgasmith commented on psi4/driver/driver.py:
```diff
@@ -689,23 +699,19 @@ def gradient(name, **kwargs):
-                energy, lowername, moleculeclone, displacement, n, ndisp, write_orbitals=False, **kwargs)
-
-        wfn = _process_displacement(energy, lowername, moleculeclone, findif_meta_dict["reference"], ndisp, ndisp,
+        wfn = _process_displacement(energy, lowername, molecule, findif_meta_dict["reference"], 1, ndisp,
```
Can you elaborate on the new looping structure here?

JonathonMisiewicz replied:
This isn't a new looping structure, just a change in the order in which we compute displacements. The reference displacement, previously done last, is now moved to be first. The argument to `_process_displacement` specifying the displacement number for print purposes has been changed accordingly.

The point of the change is so the reference is guaranteed to be available for use as an orbital guess, per https://github.com/psi4/psi4/issues/1296#issuecomment-431018568.

JonathonMisiewicz commented:
@andysim I checked DOCC compatibility just before I submitted this, and it worked. I'll add a simple DOCC set to the findif test when I fix whatever is causing tests to fail.

It'll likely be a few days before I have the time to get those fixed.

JonathonMisiewicz commented:
* Fixed mp2-module failure. Changed the symmetry reset to only occur if the user provided a default symmetry. My changes overrode Bozkaya's DF code running computations in C1, and the computational symmetry change resulted in BH3 landing on different UHF solutions.
* Fixed pywrap-all failure. Per discussion in Slack, findif now saves and resets the core variables from the reference, after all displacements are finished.
* Addressed concerns of @andysim. I now set DOCC explicitly for my test case that you can set symmetry.

JonathonMisiewicz commented:
I know the code used `P::e` before I touched it, but I don't think any of the code I added uses `P::e`. If that's not the case, let me know, and I'll see if I can remove the offending code, to help eliminate globals.

I'm holding off on documentation update until I know if I need to make other changes while I'm at it.

loriab commented:
The driver portion you're touching with findif is reactionary. It can't modernize until the c-side and proc.py improves. So you're good.

JonathonMisiewicz commented:
And, warning against modifying molecule in `_process_displacement` added.
